synedrion/src/paillier/rsa.rs

```diff
@@ -0,0 +1,282 @@
+        SecretBox::init_with(|| self.p_signed().expose_secret().into_wide())
+    }
+
+    pub fn totient(&self) -> &SecretBox<P::Uint> {
+        &self.totient
+    }
+
+    pub fn totient_bounded(&self) -> SecretBox<Bounded<P::Uint>> {
+        SecretBox::init_with(|| {
+            Bounded::new(*self.totient.expose_secret(), P::MODULUS_BITS).expect("`P::MODULUS_BITS` is valid")
+        })
+    }
+
+    pub fn totient_wide_bounded(&self) -> SecretBox<Bounded<P::WideUint>> {
+        SecretBox::init_with(|| self.totient_bounded().expose_secret().into_wide())
+    }
+
+    pub fn totient_nonzero(&self) -> SecretBox<NonZero<P::Uint>> {
+        SecretBox::init_with(|| {
+            NonZero::new(*self.totient.expose_secret()).expect(concat![
+                "φ(n) is never zero for n >= 1; n is strictly greater than 1 ",
+                "because it is (p-1)(q-1) and given that both p and q are prime ",
+                "they are both strictly greater than 1"
+            ])
+        })
+    }
+
+    /// Returns a random in range `[0, \phi(N))`.
+    pub fn random_field_elem(&self, rng: &mut impl CryptoRngCore) -> Bounded<P::Uint> {
+        Bounded::new(
+            P::Uint::random_mod(rng, self.totient_nonzero().expose_secret()),
+            P::MODULUS_BITS,
+        )
+        .expect(concat![
+            "the totient is smaller than the modulus, ",
+            "and thefore can be bounded by 2^MODULUS_BITS"
+        ])
+    }
+}
+
+/// The minimized structure containing the public RSA modulus.
+///
+/// Suitable for serialization or transmission.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) struct PublicModulusWire<P: PaillierParams>(P::Uint);
+
+impl<P: PaillierParams> PublicModulusWire<P> {
+    fn new(primes: &SecretPrimesWire<P>) -> Self {
+        Self(primes.p.expose_secret().mul_wide(primes.q.expose_secret()))
+    }
+
+    pub fn into_precomputed(self) -> PublicModulus<P> {
+        PublicModulus::new(self)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct PublicModulus<P: PaillierParams> {
+    /// The base RSA modulus $N$.
+    modulus: PublicModulusWire<P>,
+    /// Montgomery representation parameters for modulo $N$.
+    monty_params_mod_n: <P::UintMod as Monty>::Params,
+}
+
+impl<P: PaillierParams> PartialEq for PublicModulus<P> {
+    fn eq(&self, other: &PublicModulus<P>) -> bool {
+        // Only compare the moduli themselves, not the precomputed constants.
+        self.modulus.eq(&other.modulus)
+    }
+}
+
+impl<P: PaillierParams> Eq for PublicModulus<P> {}
+
+impl<P: PaillierParams> PublicModulus<P> {
+    pub fn new(modulus: PublicModulusWire<P>) -> Self {
+        let odd_modulus = Odd::new(modulus.0).expect("the RSA modulus is odd");
+        let monty_params_mod_n = P::UintMod::new_params_vartime(odd_modulus);
+        Self {
+            modulus,
+            monty_params_mod_n,
+        }
+    }
+
+    pub fn to_wire(&self) -> PublicModulusWire<P> {
```

Comment:
Not sure I understand the naming here. Is this "wire" as in "the wire format", i.e. the form of this data that ends up in messages? Or?

Reply:
Yes, exactly that.